# Coding Exercises (Part 6)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Intermediate

In [ ]:
pd.options.display.float_format = 

### Exercise 6: NA Values & Summary Statistics

In [2]:
# run the cell!
import pandas as pd
import numpy as np
import seaborn as sns
sns.set()
import matplotlib.pyplot as plt
%config InlineBackened.figure_format = 'svg'
pd.options.display.float_format = '{:.2f}'.format

In [ ]:
pd.option.display.float_format = '{:.2f}'.format

In [4]:
# run the cell!
listings = pd.read_csv("listings.csv")

In [6]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   ADR TSO  IPO_Year            Sector  \
0      NaN   1999.00     Capital Goods   
1      NaN   2016.00  Basic Industries   
2      NaN       NaN        Technology   
3      NaN   2014.00       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [8]:
# run the cell
listings.info(memory_usage= 'deep' , 
             show_counts= True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6745 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   ADR TSO     140 non-null    float64
 6   IPO_Year    3105 non-null   float64
 7   Sector      5309 non-null   object 
 8   Industry    5309 non-null   object 
dtypes: float64(4), object(5)
memory usage: 2.4 MB


51. Get the __amount__ of __NA-values__ per __column__! __Fill in the gaps__! <br> How many __NA-values__ are in the __ADR TSO__ column?

In [15]:
listings.isna().any(axis = 0)

Symbol        False
Exchange      False
Name          False
Last_Price     True
Market_Cap     True
ADR TSO        True
IPO_Year       True
Sector         True
Industry       True
dtype: bool

In [27]:
listings['ADR TSO'].isnull().sum()

6712

There are ... 6712 NA-values in the ADR TSO column!

52. Get the __amount__ of __non-null values__ per column (don´t use info() method)! <br>

In [29]:
listings.isna().any(axis = 0)

Symbol        False
Exchange      False
Name          False
Last_Price     True
Market_Cap     True
ADR TSO        True
IPO_Year       True
Sector         True
Industry       True
dtype: bool

53. __Drop__ all __rows__ with __at least one NA-Value__ from listings (don´t overwrite listings!). __How many__ rows/listing are __left__?

In [39]:
listings.dropna().shape

(100, 9)

There are only ... 100 rows left.

54. __Drop__ all __columns__ with __at least one NA-Value__ from listings (don´t overwrite listings!). __How many__ columns are __left__?

In [43]:
listings.dropna(axis= 1).shape[1]

3

There are only ... 3 columns left!

55. We finally decide to __drop__ all __columns__ with __less than 3000 non-null values__. __Overwrite__ listings! __Fill in the gaps!__

In [45]:
listings.dropna(axis = 1, thresh = 3000, inplace = True)
# It will drop the column ADR TSO

In [47]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6745 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    3105 non-null   float64
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(3), object(5)
memory usage: 428.4+ KB


56. __Filter__ listings for all rows/listings with __NA-values__ in the __Last_Price__ column. Save the new DataFrame in the variable __no_price__. <br>
__Fill in the gaps__!

In [58]:
listings.Last_Price.isna()

True

In [60]:
no_price = listings[listings.Last_Price.isna()]
no_price

Symbol Exchange                                       Name  \
151      AGM.A     NYSE  Federal Agricultural Mortgage Corporation   
190     AIG.WS     NYSE         American International Group, Inc.   
224      AKO.A     NYSE                  Embotelladora Andina S.A.   
225      AKO.B     NYSE                  Embotelladora Andina S.A.   
254       ALIT   NASDAQ                                Alight Inc.   
...        ...      ...                                        ...   
6170    TRNE.U     NYSE                    Trine Acquisition Corp.   
6403   UUUU.WS     AMEX                           Energy Fuels Inc   
6551  VST.WS.A     NYSE                        Vistra Energy Corp.   
6717     WSO.B     NYSE                               Watsco, Inc.   
6839     ZNWAA   NASDAQ                         Zion Oil & Gas Inc   

      Last_Price  Market_Cap  IPO_Year  Sector              Industry  
151          NaN        0.00       NaN     NaN                   NaN  
190          NaN        0.00       NaN     NaN                   NaN  
224          NaN        0.00       NaN     NaN                   NaN  
225          NaN        0.00       NaN     NaN                   NaN  
254          NaN        0.00       NaN     NaN                   NaN  
...          ...         ...       ...     ...                   ...  
6170         NaN        0.00   2019.00     NaN                   NaN  
6403         NaN        0.00       NaN     NaN                   NaN  
6551         NaN        0.00       NaN     NaN                   NaN  
6717         NaN        0.00       NaN     NaN                   NaN  
6839         NaN        0.00       NaN  Energy  Oil & Gas Production  

[107 rows x 8 columns]

What do all listings with __NA-values in the Last_Price__ column have in common (hint: check the __Market_Cap__ column)?

In [ ]:
# run the cell and inspect!
no_price.head()

In [ ]:
# run the cell and inspect!
no_price.info()

In [62]:
#run the cell and inspect!
no_price.Market_Cap.value_counts()

Market_Cap
0.00    107
Name: count, dtype: int64

The Market_Cap is 0.0 for all listings with NA-values in the Last_Price column. Therefore, it might make sense to...

57. ...__Fill all NA-Values__ in the __Last_Price__ column of listings with __0.0__! __Save the change__!

In [65]:
no_price.Last_Price.fillna(value = 0 , inplace= True)
no_price

C:\Users\ahmad\AppData\Local\Temp\ipykernel_30748\458624666.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  no_price.Last_Price.fillna(value = 0 , inplace= True)


Symbol Exchange                                       Name  \
151      AGM.A     NYSE  Federal Agricultural Mortgage Corporation   
190     AIG.WS     NYSE         American International Group, Inc.   
224      AKO.A     NYSE                  Embotelladora Andina S.A.   
225      AKO.B     NYSE                  Embotelladora Andina S.A.   
254       ALIT   NASDAQ                                Alight Inc.   
...        ...      ...                                        ...   
6170    TRNE.U     NYSE                    Trine Acquisition Corp.   
6403   UUUU.WS     AMEX                           Energy Fuels Inc   
6551  VST.WS.A     NYSE                        Vistra Energy Corp.   
6717     WSO.B     NYSE                               Watsco, Inc.   
6839     ZNWAA   NASDAQ                         Zion Oil & Gas Inc   

      Last_Price  Market_Cap  IPO_Year  Sector              Industry  
151         0.00        0.00       NaN     NaN                   NaN  
190         0.00        0.00       NaN     NaN                   NaN  
224         0.00        0.00       NaN     NaN                   NaN  
225         0.00        0.00       NaN     NaN                   NaN  
254         0.00        0.00       NaN     NaN                   NaN  
...          ...         ...       ...     ...                   ...  
6170        0.00        0.00   2019.00     NaN                   NaN  
6403        0.00        0.00       NaN     NaN                   NaN  
6551        0.00        0.00       NaN     NaN                   NaN  
6717        0.00        0.00       NaN     NaN                   NaN  
6839        0.00        0.00       NaN  Energy  Oil & Gas Production  

[107 rows x 8 columns]

58. __Filter__ listings for all rows/listings that have __non-null values__ in the __IPO_Year__ column! __Save__ the new DataFrame in the variable __known_year__. <br> 
__Fill in the gaps!__

In [67]:
listings.IPO_Year.notnull()

0        True
1        True
2       False
3        True
4       False
        ...  
6847     True
6848     True
6849     True
6850     True
6851    False
Name: IPO_Year, Length: 6852, dtype: bool

In [71]:
known_year = listings.loc[listings.IPO_Year.notnull()]
known_year

Symbol Exchange                           Name  Last_Price  \
0         A     NYSE     Agilent Technologies, Inc.       81.68   
1        AA     NYSE              Alcoa Corporation       29.15   
3       AAC     NYSE             AAC Holdings, Inc.        2.16   
8      AAOI   NASDAQ  Applied Optoelectronics, Inc.       12.30   
11     AAPL   NASDAQ                     Apple Inc.      200.62   
...     ...      ...                            ...         ...   
6846    ZTS     NYSE                    Zoetis Inc.      101.33   
6847   ZUMZ   NASDAQ                    Zumiez Inc.       26.72   
6848    ZUO     NYSE                    Zuora, Inc.       19.79   
6849   ZYME     NYSE                 Zymeworks Inc.       15.74   
6850   ZYNE   NASDAQ  Zynerba Pharmaceuticals, Inc.        7.85   

          Market_Cap  IPO_Year             Sector  \
0     25934696179.92   1999.00      Capital Goods   
1      5407809589.60   2016.00   Basic Industries   
3        53141086.80   2014.00        Health Care   
8       244556139.90   2013.00         Technology   
11   945979473600.00   1980.00         Technology   
...              ...       ...                ...   
6846  48552001684.81   2013.00        Health Care   
6847    681919062.56   2005.00  Consumer Services   
6848   2147266454.00   2018.00         Technology   
6849    504078206.26   2017.00        Health Care   
6850    165399468.60   2015.00        Health Care   

                                              Industry  
0     Biotechnology: Laboratory Analytical Instruments  
1                                             Aluminum  
3                                 Medical Specialities  
8                                       Semiconductors  
11                              Computer Manufacturing  
...                                                ...  
6846                             Major Pharmaceuticals  
6847                    Clothing/Shoe/Accessory Stores  
6848           Computer Software: Prepackaged Software  
6849                             Major Pharmaceuticals  
6850                             Major Pharmaceuticals  

[3105 rows x 8 columns]

__How many__ listings with known IPO_Year are in the __known_year__ DataFrame?

In [81]:
# run the cell!
known_year.head()

Symbol Exchange                           Name  Last_Price      Market_Cap  \
0       A     NYSE     Agilent Technologies, Inc.       81.68  25934696179.92   
1      AA     NYSE              Alcoa Corporation       29.15   5407809589.60   
3     AAC     NYSE             AAC Holdings, Inc.        2.16     53141086.80   
8    AAOI   NASDAQ  Applied Optoelectronics, Inc.       12.30    244556139.90   
11   AAPL   NASDAQ                     Apple Inc.      200.62 945979473600.00   

    IPO_Year            Sector  \
0    1999.00     Capital Goods   
1    2016.00  Basic Industries   
3    2014.00       Health Care   
8    2013.00        Technology   
11   1980.00        Technology   

                                            Industry  
0   Biotechnology: Laboratory Analytical Instruments  
1                                           Aluminum  
3                               Medical Specialities  
8                                     Semiconductors  
11                            Computer Manufacturing

In [83]:
# run the cell!
known_year.shape

(3105, 8)

There are ... 3105 listings with known IPO_Year!

Finally, we figured out that __IPO_Year__ information is __not available__ for stocks with an IPO __prior to 1970__. Therefore...

In [102]:
known_year.agg({'Last_Price':['min','max'],
                'IPO_Year':['min','max']})

Last_Price  IPO_Year
min        0.00   1972.00
max     4299.99   2019.00

In [106]:
known_year.sort_values(by = ['IPO_Year'] , ascending= True)

Symbol Exchange                          Name  Last_Price  \
1439    COKE   NASDAQ  Coca-Cola Consolidated, Inc.      294.70   
294     AMAT   NASDAQ       Applied Materials, Inc.       42.17   
6612    WDFC   NASDAQ                 WD-40 Company      163.60   
3539    KLAC   NASDAQ        KLA-Tencor Corporation      123.34   
11      AAPL   NASDAQ                    Apple Inc.      200.62   
...      ...      ...                           ...         ...   
356    ANDAR   NASDAQ  Andina Acquisition Corp. III        0.33   
357    ANDAU   NASDAQ  Andina Acquisition Corp. III       10.20   
358    ANDAW   NASDAQ  Andina Acquisition Corp. III        0.18   
3021    HOTH   NASDAQ       Hoth Therapeutics, Inc.        5.39   
6170  TRNE.U     NYSE       Trine Acquisition Corp.         NaN   

          Market_Cap  IPO_Year                 Sector  \
1439   2104584430.90   1972.00  Consumer Non-Durables   
294   40035886321.53   1972.00             Technology   
6612   2257237298.40   1973.00       Basic Industries   
3539  20186316279.92   1980.00          Capital Goods   
11   945979473600.00   1980.00             Technology   
...              ...       ...                    ...   
356              NaN   2019.00                Finance   
357              NaN   2019.00                Finance   
358              NaN   2019.00                Finance   
3021     50805945.96   2019.00            Health Care   
6170            0.00   2019.00                    NaN   

                                 Industry  
1439  Beverages (Production/Distribution)  
294                        Semiconductors  
6612                      Major Chemicals  
3539                Electronic Components  
11                 Computer Manufacturing  
...                                   ...  
356                     Business Services  
357                     Business Services  
358                     Business Services  
3021                Major Pharmaceuticals  
6170                                  NaN  

[3105 rows x 8 columns]

59. ... we decide to __fill all NA-Values__ in the __IPO_Year__ column with __1969__. __Save__ the change!

In [112]:
known_year.IPO_Year.fillna(1969 , inplace= True)

C:\Users\ahmad\AppData\Local\Temp\ipykernel_30748\4022326382.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  known_year.IPO_Year.fillna(1969 , inplace= True)


In [114]:
known_year

Symbol Exchange                           Name  Last_Price  \
0         A     NYSE     Agilent Technologies, Inc.       81.68   
1        AA     NYSE              Alcoa Corporation       29.15   
3       AAC     NYSE             AAC Holdings, Inc.        2.16   
8      AAOI   NASDAQ  Applied Optoelectronics, Inc.       12.30   
11     AAPL   NASDAQ                     Apple Inc.      200.62   
...     ...      ...                            ...         ...   
6846    ZTS     NYSE                    Zoetis Inc.      101.33   
6847   ZUMZ   NASDAQ                    Zumiez Inc.       26.72   
6848    ZUO     NYSE                    Zuora, Inc.       19.79   
6849   ZYME     NYSE                 Zymeworks Inc.       15.74   
6850   ZYNE   NASDAQ  Zynerba Pharmaceuticals, Inc.        7.85   

          Market_Cap  IPO_Year             Sector  \
0     25934696179.92   1999.00      Capital Goods   
1      5407809589.60   2016.00   Basic Industries   
3        53141086.80   2014.00        Health Care   
8       244556139.90   2013.00         Technology   
11   945979473600.00   1980.00         Technology   
...              ...       ...                ...   
6846  48552001684.81   2013.00        Health Care   
6847    681919062.56   2005.00  Consumer Services   
6848   2147266454.00   2018.00         Technology   
6849    504078206.26   2017.00        Health Care   
6850    165399468.60   2015.00        Health Care   

                                              Industry  
0     Biotechnology: Laboratory Analytical Instruments  
1                                             Aluminum  
3                                 Medical Specialities  
8                                       Semiconductors  
11                              Computer Manufacturing  
...                                                ...  
6846                             Major Pharmaceuticals  
6847                    Clothing/Shoe/Accessory Stores  
6848           Computer Software: Prepackaged Software  
6849                             Major Pharmaceuticals  
6850                             Major Pharmaceuticals  

[3105 rows x 8 columns]

Make a __final inspection__ of our listings DataFrame. Which __columns__ still contain __NA-values__?

In [108]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0   1999.00     Capital Goods   
1   2016.00  Basic Industries   
2       NaN        Technology   
3   2014.00       Health Care   
4       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [122]:
listings.Last_Price.fillna(value=0 ,inplace= True)
listings.IPO_Year.fillna(value = 1969 , inplace = True)
listings.Market_Cap.fillna(value = 0 , inplace = True)

In [124]:
#run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  6852 non-null   float64
 5   IPO_Year    6852 non-null   float64
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(3), object(5)
memory usage: 428.4+ KB


The columns ... Sector & Industry still contain NA-values!

60. __Transform the datatype__ in the IPO_Year column to integer (beyond the scope of this course, run the cell!)

In [126]:
#run the cell!
listings.IPO_Year = listings.IPO_Year.astype("int")

In [130]:
#run the cell and inspect!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  6852 non-null   float64
 5   IPO_Year    6852 non-null   int32  
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(2), int32(1), object(5)
memory usage: 401.6+ KB


In [140]:
listings.nunique()

Symbol        6840
Exchange         3
Name          5766
Last_Price    4471
Market_Cap    5835
IPO_Year        43
Sector          12
Industry       135
dtype: int64

In [150]:
listings.Sector = listings.Sector.astype('category')
listings.Industry = listings.Industry.astype('category')

In [152]:
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   Symbol      6852 non-null   object  
 1   Exchange    6852 non-null   object  
 2   Name        6852 non-null   object  
 3   Last_Price  6852 non-null   float64 
 4   Market_Cap  6852 non-null   float64 
 5   IPO_Year    6852 non-null   int32   
 6   Sector      5309 non-null   category
 7   Industry    5309 non-null   category
dtypes: category(2), float64(2), int32(1), object(3)
memory usage: 320.1+ KB


In [ ]:
#run the cell!
listings.head()

61. __Save and export__ the cleaned DataFrame to the new csv-file __listings_clean.csv__! __Fill in the gaps!__

In [154]:
listings.to_csv("listings_clean.csv", index = False)

62. __Re-Import__ listings_clean.csv and save the DataFrame in the variable __listings_clean__!

In [156]:
listings_clean = pd.read_csv("listings_clean.csv")

In [158]:
# run the cell!
listings_clean.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0      1999     Capital Goods   
1      2016  Basic Industries   
2      1969        Technology   
3      2014       Health Care   
4      1969    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [160]:
# run the cell!
listings_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  6852 non-null   float64
 5   IPO_Year    6852 non-null   int64  
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 428.4+ KB


63. Call an appropriate method on the __listings_clean__ DataFrame that provides plenty of __summary statistics__!

In [162]:
listings_clean.describe()

Last_Price      Market_Cap  IPO_Year
count     6852.00         6852.00   6852.00
mean        36.79   6234182291.95   1987.14
std         98.58  32951998468.28     20.89
min          0.00            0.00   1969.00
25%          7.70     44072156.58   1969.00
50%         19.98    354821838.54   1969.00
75%         36.85   2195527462.22   2012.00
max       4299.99 945979473600.00   2019.00

63a. __Calculate__ the __mean__ and the __standard deviation__ for the numerical columns (use the appropriate method!).

In [170]:
listings_clean.drop('IPO_Year', axis = 1).select_dtypes(include= 'number').agg(['mean' , 'std'])

Last_Price     Market_Cap
mean       36.79  6234182291.95
std        98.58 32951998468.28

64. Get some summary statistics on all __non-numerical columns__ by passing __"O"__ to the __include parameter__ of the __describe()__ method! __Fill in the gaps!__ <br>
What is the __most frequent__ value in the __Industry__ column? 

In [172]:
listings_clean.describe(include = "O")

Symbol Exchange            Name   Sector               Industry
count    6852     6852            6852     5309                   5309
unique   6840        3            5766       12                    135
top      CDOR   NASDAQ  Public Storage  Finance  Major Pharmaceuticals
freq        2     3435              14     1059                    470

The most frequent Industry is ... Major Pharmaceuticals with 470 appearances.

65. Get the __maximum__ values for all columns! Which is the (alphabetically) __last Symbol__?

In [174]:
listings_clean.Symbol.max()

'ZYXI'

The (alphabetically) last Symbol is ... ZYXI!

66. __Calculate__ the __total Market_Cap__ over all listings.

In [176]:
listings_clean.Market_Cap.sum()

42716617064473.484

67. Create a __correlation matrix__ for the __listings_clean__ DataFrame! Are Market_Cap and Last_Price __positively correlated__?

In [204]:
(
    listings_clean.
    drop('IPO_Year', axis = 1).
    corr(numeric_only= True).
    style.
    background_gradient(cmap = "seismic")
)

Market_Cap and Last_Price are ... positively correlated.The higher the Price, the higher the Market Capitalization.

# Well Done!

------------------------------

# Hints (Spoiler!)

51. isna() method, sum() method

52. notna() method, sum() method

53. dropna(axis = 0) method and shape attribute

54. dropna(axis = 1) method and shape attribute

55. dropna() method, thresh parameter

56. pass listings.Last_Price.isna() to the loc operator

57. fillna() method

58. pass listings.IPO_Year.notna() to the loc operator

59. fillna() method

60. run the cell!

61. to_csv() method

62. You should know this already!

63. describe() method 

63a. agg() method, pass list with "mean" and "std"

64. describe() method and include parameter

65. max() method

66. sum() method

67. corr() method